File: server/connect/mqtt.js
const mqtt = require("mqtt");
const State = require("../models/State");
const Action = require("../models/Action");

const broker = "mqtt://172.6.0.240";
const port = 1883;
const topic_sensor = "sensor/data";
const topic_control = "control/movement";

const client = mqtt.connect(broker, { port });

client.on("connect", () => {
  console.log("Connected to MQTT broker");

  client.subscribe([topic_sensor, topic_control], (err) => {
    if (err) {
      console.error("Subscription error:", err);
    } else {
      console.log(`Subscribed to topics: ${topic_sensor}, ${topic_control}`);
    }
  });
});

client.on("message", (topic, message) => {
  console.log(`Received message: ${message} on topic: ${topic}`);

  //错误处理
  if (topic === topic_sensor) {
    let data;
    try {
      data = JSON.parse(message);
    } catch (e) {
      console.error("Failed to parse message as JSON:", message);
      return;
    }

    //解析data
    const { temperature, pressure, depth } = data;
    const time = new Date();
    const newState = new State({ temperature, pressure, depth, time });

    newState
      .save()
      .then(() => {
        console.log("State saved successfully");
      })
      .catch((error) => {
        console.error("Error saving state", error);
      });
  } else if (topic === topic_control) {
    console.log("Control signal received:", message.toString());
  }
});

// 向控制主题发送命令的函数
function sendCommand(command) {
  client.publish(topic_control, command, (err) => {
    if (err) {
      console.error("Failed to send control command:", err);
    } else {
      console.log("Control command sent successfully:", command);
    }

    const time = new Date();
    const newAction = new Action({ command, time });

    newAction
      .save()
      .then(() => {
        console.log("Action saved successfully");
      })
      .catch((error) => {
        console.error("Error saving action", error);
      });
  });
}

module.exports = { client, sendCommand };

File: mqtt_broker/mqtt.py
import paho.mqtt.client as mqtt
import random
import time
import threading
import json

# MQTT服务器设置
broker = "172.6.0.240"  # MQTT代理地址
port = 1883  # MQTT代理端口
topic_sensor = "sensor/data"
topic_control = "control/movement"

# 传感器值
temperature = 0.0
pressure = 0.0
depth = 0.0


def publish_sensor_data():
    global temperature, pressure, depth
    client = mqtt.Client()
    client.connect(broker, port)

    while True:
        # 模拟传感器数据
        temperature = round(random.uniform(15.0, 30.0), 2)  # 温度范围 15-30℃
        pressure = round(random.uniform(95.0, 105.0), 2)  # 气压范围 95-105 KPa
        depth = round(random.uniform(0.0, 10.0), 2)  # 深度范围 0-10 M

        # 使用 json.dumps 生成 JSON 格式的payload
        payload = json.dumps({
            "temperature": temperature,
            "pressure": pressure,
            "depth": depth
        })

        client.publish(topic_sensor, payload)
        print(f"Published: {payload}")
        time.sleep(10)


def on_message(client, userdata, msg):
    print(f"Received message: {msg.payload.decode()} on topic {msg.topic}")


def subscribe_control_signals():
    client = mqtt.Client()
    client.on_message = on_message
    client.connect(broker, port)
    client.subscribe(topic_control)
    client.loop_start()  # 启动循环以接收消息

    while True:
        time.sleep(1)  # 等待接收信号


if __name__ == "__main__":
    sensor_thread = threading.Thread(target=publish_sensor_data)
    sensor_thread.start()
    subscribe_control_signals()
